contrib/scalar/scalar.c

```diff
@@ -346,9 +348,181 @@ static int cmd_clone(int argc, const char **argv)
+			continue;
+
+		strbuf_setlen(&buf, len);
+		strbuf_addstr(&buf, e->d_name);
+
+		if ((e->d_type == DT_REG && stage_file(git_dir, buf.buf)) ||
+		    (e->d_type == DT_DIR && recurse &&
+		     stage_directory(git_dir, buf.buf, recurse)))
+			res = -1;
+	}
+
+	closedir(dir);
+	strbuf_release(&buf);
+	return res;
+}
+
+static int index_to_zip(const char *git_dir)
+{
+	struct child_process cp = CHILD_PROCESS_INIT;
+	struct strbuf oid = STRBUF_INIT;
+
+	cp.git_cmd = 1;
+	strvec_pushl(&cp.args, "--git-dir", git_dir, "write-tree", NULL);
+	if (pipe_command(&cp, NULL, 0, &oid, the_hash_algo->hexsz + 1,
+			 NULL, 0))
+		return error(_("could not write temporary tree object"));
+
+	strbuf_rtrim(&oid);
+	child_process_init(&cp);
+	cp.git_cmd = 1;
+	strvec_pushl(&cp.args, "--git-dir", git_dir, "archive", "-o", NULL);
+	strvec_pushf(&cp.args, "%s.zip", git_dir);
+	strvec_pushl(&cp.args, oid.buf, "--", NULL);
+	strbuf_release(&oid);
+	return run_command(&cp);
+}
+
 static int cmd_diagnose(int argc, const char **argv)
 {
-	die(N_("'%s' not yet implemented"), argv[0]);
+	struct strbuf tmp_dir = STRBUF_INIT;
+	time_t now = time(NULL);
+	struct tm tm;
+	struct strbuf path = STRBUF_INIT, buf = STRBUF_INIT;
+	int res = 0;
+
+	if (argc != 1)
+		die("'scalar diagnose' does not accept any arguments");
+
+	strbuf_addstr(&tmp_dir, ".scalarDiagnostics/scalar_");
+	strbuf_addftime(&tmp_dir, "%Y%m%d_%H%M%S",
+			localtime_r(&now, &tm), 0, 0);
+	if (run_git(NULL, "init", "-q", "-b", "dummy",
+		    "--bare", tmp_dir.buf, NULL)) {
+		res = error(_("could not initialize temporary repository"));
+		goto diagnose_cleanup;
+	}
+
+	strbuf_reset(&buf);
+	strbuf_addf(&buf, "Collecting diagnostic info into temp folder %s\n\n",
+		    tmp_dir.buf);
+
+	strbuf_addf(&buf, "git version %s\n", git_version_string);
+	strbuf_addf(&buf, "built from commit: %s\n\n",
+		    git_built_from_commit_string[0] ?
+		    git_built_from_commit_string : "(n/a)");
+
+	strbuf_addf(&buf, "Enlistment root: %s\n", the_repository->worktree);
+	strbuf_addf(&buf,
+		    "Cache Server: None\n"
+		    "Local Cache:\n"
+		    "\n"
+		    "TODO: acquire disk space information\n");
+	fwrite(buf.buf, buf.len, 1, stdout);
+
+	if ((res = stage(tmp_dir.buf, &buf, "diagnostics.log")))
+		goto diagnose_cleanup;
+
+	if ((res = stage_directory(tmp_dir.buf, ".git", 0)) ||
+	    (res = stage_directory(tmp_dir.buf, ".git/hooks", 0)) ||
+	    (res = stage_directory(tmp_dir.buf, ".git/info", 0)) ||
+	    (res = stage_directory(tmp_dir.buf, ".git/logs", 1)) ||
+	    (res = stage_directory(tmp_dir.buf, ".git/objects/info", 0)))
+		goto diagnose_cleanup;
+
+	/*
+	 * TODO: add more stuff:
+	 * disk space info
+	 * LogDirectoryEnumeration(...DotGit.Objects.Root), ScalarConstants.DotGit.Objects.Pack.Root, "packs-local.txt");
+	 * LogLooseObjectCount(...DotGit.Objects.Root), ScalarConstants.DotGit.Objects.Root, "objects-local.txt");
+	 *
+	 * CopyLocalCacheData(archiveFolderPath, gitObjectsRoot);
+	 */
```

Comment:
Thanks for commenting here. These are my favorite things to inspect in diagnose logs. We should be able to collect this data for each odb. It doesn't need to use the same structure as before.

Reply:
Right, I figured. They were also the least trivial to implement, that's why I left them for later ;-)